src/main/java/org/elasticsearch/indices/SyncedFlushService.java

```diff
@@ -0,0 +1,755 @@
+                            return new SyncedFlushResponse();
+                        }
+
+                        @Override
+                        public void handleResponse(SyncedFlushResponse response) {
+                            SyncedFlushResponse existing = results.put(shard, response);
+                            assert existing == null : "got two answers for node [" + node + "]";
+                            // count after the assert so we won't decrement twice in handleException
+                            if (countDownLatch.countDown()) {
+                                listener.onResponse(new SyncedFlushResult(shardId, syncId, results));
+                            }
+                        }
+
+                        @Override
+                        public void handleException(TransportException exp) {
+                            logger.trace("{} error while performing synced flush on [{}], skipping", exp, shardId, shard);
+                            results.put(shard, new SyncedFlushResponse(exp.getMessage()));
+                            if (countDownLatch.countDown()) {
+                                listener.onResponse(new SyncedFlushResult(shardId, syncId, results));
+                            }
+                        }
+
+                        @Override
+                        public String executor() {
+                            return ThreadPool.Names.SAME;
+                        }
+                    });
+        }
+
+    }
+
+    /**
+     * send presync requests to all started copies of the given shard
+     */
+    Map<String, byte[]> sendPreSyncRequests(final List<ShardRouting> shards, final ClusterState state, final ShardId shardId) {
+        final CountDownLatch countDownLatch = new CountDownLatch(shards.size());
+        final Map<String, byte[]> commitIds = ConcurrentCollections.newConcurrentMap();
+        for (final ShardRouting shard : shards) {
+            logger.trace("{} sending pre-synced flush request to {}", shardId, shard);
+            final DiscoveryNode node = state.nodes().get(shard.currentNodeId());
+            if (node == null) {
+                logger.trace("{} shard routing {} refers to an unknown node. skipping.", shardId, shard);
+                countDownLatch.countDown();
+                continue;
+            }
+            transportService.sendRequest(node, PRE_SYNCED_FLUSH_ACTION_NAME, new PreSyncedFlushRequest(shard.shardId()), new BaseTransportResponseHandler<PreSyncedFlushResponse>() {
+                @Override
+                public PreSyncedFlushResponse newInstance() {
+                    return new PreSyncedFlushResponse();
+                }
+
+                @Override
+                public void handleResponse(PreSyncedFlushResponse response) {
+                    byte[] existing = commitIds.put(node.id(), response.commitId());
+                    assert existing == null : "got two answers for node [" + node + "]";
```

Comment:
actually, we have to put the assert before. In case it fails we call handleFailure and therefore would decrement twice if we first decrement and then assert (although I really think if this assertion fails we have a much worse problem.)
There is a comment in the code just beneath this line.